src/NServiceBus.Core/Transports/Msmq/ReceiveWithTransactionScope.cs

```diff
@@ -19,45 +19,118 @@ public ReceiveWithTransactionScope(TransactionOptions transactionOptions)
+            Message message = null;
 
-                try
-                {
-                    headers = MsmqUtilities.ExtractHeaders(message);
-                }
-                catch (Exception ex)
+            try
+            {
+                using (var scope = new TransactionScope(TransactionScopeOption.Required, transactionOptions, TransactionScopeAsyncFlowOption.Enabled))
                 {
-                    var error = $"Message '{message.Id}' is corrupt and will be moved to '{errorQueue.QueueName}'";
-                    Logger.Error(error, ex);
+                    message = inputQueue.Receive(TimeSpan.FromMilliseconds(10), MessageQueueTransactionType.Automatic);
 
-                    errorQueue.Send(message, MessageQueueTransactionType.Automatic);
+                    if (PerformRecoverabilityAction(message))
+                    {
+                        scope.Complete();
+                        return;
+                    }
 
-                    scope.Complete();
-                    return;
-                }
+                    try
+                    {
+                        headers = MsmqUtilities.ExtractHeaders(message);
+                    }
+                    catch (Exception ex)
+                    {
+                        var error = $"Message '{message.Id}' is corrupt and will be moved to '{errorQueue.QueueName}'";
+                        Logger.Error(error, ex);
 
-                using (var bodyStream = message.BodyStream)
-                {
-                    var ambientTransaction = new TransportTransaction();
-                    ambientTransaction.Set(Transaction.Current);
-                    var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
+                        errorQueue.Send(message, MessageQueueTransactionType.Automatic);
 
-                    await onMessage(pushContext).ConfigureAwait(false);
-                }
+                        scope.Complete();
+                        return;
+                    }
+
+                    using (var bodyStream = message.BodyStream)
+                    {
+                        var ambientTransaction = new TransportTransaction();
+                        ambientTransaction.Set(Transaction.Current);
+                        var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
 
-                if (!cancellationTokenSource.Token.IsCancellationRequested)
+                        await onMessage(pushContext).ConfigureAwait(false);
+                    }
+
+                    if (!cancellationTokenSource.Token.IsCancellationRequested)
+                    {
+                        scope.Complete();
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                if (message == null)
```

Comment:
Note that the pump will deal with some of the cases where this is expected:

https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/Msmq/MessagePump.cs#L158